kuma/core/middleware.py

```diff
@@ -1,93 +1,181 @@
 import contextlib
-import urllib
 from urlparse import urljoin
 
 from django.conf import settings
 from django.contrib.sessions.middleware import SessionMiddleware
-from django.core.urlresolvers import resolve, Resolver404
+from django.core.urlresolvers import get_script_prefix, resolve, Resolver404
 from django.http import (HttpResponseForbidden,
                          HttpResponsePermanentRedirect,
                          HttpResponseRedirect)
 from django.utils import translation
 from django.utils.encoding import iri_to_uri, smart_str
+from django.utils.six.moves.urllib.parse import (urlencode, urlsplit,
+                                                 urlunsplit)
 from whitenoise.middleware import WhiteNoiseMiddleware
 
 from kuma.wiki.views.legacy import (mindtouch_to_kuma_redirect,
                                     mindtouch_to_kuma_url)
 
 from .decorators import add_shared_cache_control
-from .urlresolvers import Prefixer, set_url_prefixer, split_path
-from .utils import is_untrusted, urlparams
+from .i18n import (get_language,
+                   get_language_from_path,
+                   get_language_from_request)
+from .utils import is_untrusted
 from .views import handler403
 
 
-class LocaleURLMiddleware(object):
+class LangSelectorMiddleware(object):
     """
-    Based on zamboni.amo.middleware.
-    Tried to use localeurl but it choked on 'en-US' with capital letters.
+    Redirect requests with a ?lang= parameter.
 
-    1. Search for the locale.
-    2. Save it in the request.
-    3. Strip them from the URL.
+    This should appear higher than LocaleMiddleware in the middleware list.
     """
 
     def process_request(self, request):
-        prefixer = Prefixer(request)
-        set_url_prefixer(prefixer)
-        full_path = prefixer.fix(prefixer.shortened_path)
-        lang = request.GET.get('lang')
-
-        if lang in dict(settings.LANGUAGES):
-            # Blank out the locale so that we can set a new one. Remove lang
-            # from the query params so we don't have an infinite loop.
-            prefixer.locale = ''
-            new_path = prefixer.fix(prefixer.shortened_path)
-            query = dict((smart_str(k), v) for
+        """Redirect if ?lang query parameter is valid."""
+        query_lang = request.GET.get('lang')
+        if query_lang not in dict(settings.LANGUAGES):
```

Comment:
Nit. I was thinking that since this is run for almost every request, and that most requests will not have a `lang` query parameter, we could short circuit the need for a dictionary lookup with:
```python
if not (query_lang and (query_lang in dict(settings.LANGUAGES))):
```

Reply:
We have a few cases where we'd like a quick check of the supported Kuma language code, so I'll go a little farther down this path and avoid creating a ``dict`` every time.